### PR TITLE
Fix flaky testParquetTimestampPredicatePushdown

### LINKED ISSUE #4314: Flaky test TestHiveIntegrationSmokeTest.testParquetTimestampPredicatePushdown
```2020-07-01T22:33:57.6602506Z [ERROR] Tests run: 1506, Failures: 1, Errors: 0, Skipped: 72, Time elapsed: 1,774.171 s <<< FAILURE! - in TestSuite
2020-07-01T22:33:57.6609950Z [ERROR] testParquetTimestampPredicatePushdown(io.prestosql.plugin.hive.TestHiveIntegrationSmokeTest)  Time elapsed: 0.433 s  <<< FAILURE!
2020-07-01T22:33:57.6624879Z java.lang.AssertionError: 
2020-07-01T22:33:57.6625170Z 
2020-07-01T22:33:57.6626139Z Expecting:
2020-07-01T22:33:57.6626295Z  <0L>
2020-07-01T22:33:57.6627381Z to be greater than:
2020-07-01T22:33:57.6627791Z  <0L> 
2020-07-01T22:33:57.6635725Z 	at io.prestosql.plugin.hive.TestHiveIntegrationSmokeTest.testParquetTimestampPredicatePushdown(TestHiveIntegrationSmokeTest.java:4135)
2020-07-01T22:33:57.6640344Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2020-07-01T22:33:57.6645766Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2020-07-01T22:33:57.6651629Z 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2020-07-01T22:33:57.6662721Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2020-07-01T22:33:57.6666980Z 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
2020-07-01T22:33:57.6669659Z 	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
2020-07-01T22:33:57.6672773Z 	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
2020-07-01T22:33:57.6676372Z 	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
2020-07-01T22:33:57.6748641Z 	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
2020-07-01T22:33:57.6752787Z 	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
2020-07-01T22:33:57.6757887Z 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
2020-07-01T22:33:57.6762828Z 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
2020-07-01T22:33:57.6765611Z 	at java.base/java.lang.Thread.run(Thread.java:834)
```

https://github.com/prestosql/presto/pull/4224/checks?check_run_id=828279182

### PR DESCRIPTION
**Pull request by MiguelWeezardo**
Query info contains zeroes if updates haven't been applied yet.
Fixed by retrying with increasing delay between query and getQueryInfo.

This fixes #4314

### CODE REVIEW

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 1]
```diff
@@ -7056,4 +7060,60 @@ public TypeAndEstimate(Type type, EstimatedStatsAndCost estimate)
...
+        }
+
+        private long checkProcessedBytes()
+                throws Exception
+        {
+            try {
+                assertThat(processedInputBytes).isGreaterThan(0);
+            }
+            catch (AssertionError error) {
+                // If query statistics haven't been updated at the time we issued getQueryInfo(), we will mistakenly see 0 processedInputBytes.
```
Comment (losipiuk) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:7097]: can't we just retry calling `getQueryInfo` once again in a while?

Wouldn't simple assertEventually work here?

Comment (MiguelWeezardo) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:7097]: When I've investigated similar problems in the past it seemed that sometimes query stats updates get applied seconds after the query finishes, an sometimes they get lost in transfer. I suspect TCP retransmissions to be the cause, but I never spent enough time to confirm this. 

This is why the query itself also needs to be resubmitted - getting stats repeatedly for the same query won't help if  coordinator already finished collecting stats updates for this particular query.

As for using `assertEventually` - It would be easier to understand, but the race condition in each retry would have the same chances of happening. Thanks to RetryDriver separating `onRetry`, `run` and exponential backoff, we can increase sleep times between the query and asking for it's stats, so the race condition has less of a chance of happening.

Comment (losipiuk) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:7097]: I have hard time with this one. Retry driver does not rally play well with being uses as exponenetially growing `sleep` between two statements. The fact that we need to run query in `catch` clause. As well as in constructor for 0-seconds sleep, makes it hard to follow.

What about sth like:
```java
       ExponentialSleeper sleeper = new ExponentialSleeper(
                new Duration(0, SECONDS),
                new Duration(5, SECONDS),
                new Duration(1, SECONDS),
                1.5);
        assertEventually(new Duration(30, SECONDS), () -> {
            ResultWithQueryId<MaterializedResult> result = queryRunner.executeWithQueryId(
                    getSession(),
                    "SELECT * FROM test_parquet_timestamp_predicate_pushdown WHERE t = TIMESTAMP '2012-10-31 01:00'");
            sleeper.sleep();
            assertThat(getQueryInfo(queryRunner, result).getQueryStats().getProcessedInputDataSize().toBytes()).isGreaterThan(0);
        });
    }

    private static class ExponentialSleeper
    {
        private Duration nextSleepTime;
        private final Duration maxSleepTime;
        private final Duration minSleepIncrement;
        private final double sleepIncrementFactor;

        private ExponentialSleeper(Duration minSleepTime, Duration maxSleepTime, Duration minSleepIncrement, double sleepIncrementFactor)
        {
            this.nextSleepTime = minSleepTime;
            this.maxSleepTime = maxSleepTime;
            this.minSleepIncrement = minSleepIncrement;
            this.sleepIncrementFactor = sleepIncrementFactor;
        }

        public void sleep()
        {
            try {
                Thread.sleep(nextSleepTime.toMillis());
                long incrementMillis = (long) (nextSleepTime.toMillis() * sleepIncrementFactor - nextSleepTime.toMillis());
                if (incrementMillis < minSleepIncrement.toMillis()) {
                    incrementMillis = minSleepIncrement.toMillis();
                }
                nextSleepTime = new Duration(nextSleepTime.toMillis() + incrementMillis, MILLISECONDS);
                if (nextSleepTime.compareTo(maxSleepTime) > 0) {
                    nextSleepTime = maxSleepTime;
                }
            }
            catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                throw new RuntimeException(e);
            }
        }
    }

```

Comment (losipiuk) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:7097]: I guess this new class may come useful in other places too (maybe)

**Comment by MiguelWeezardo** [comment]
@losipiuk could you merge this? I'm not on the maintainers list yet.

**Comment by losipiuk** [comment]
> @losipiuk could you merge this?

Sure. sorry for the delay. I think it was still going through CI when I last looked at it.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 2]
```diff
@@ -7056,4 +7069,39 @@ public TypeAndEstimate(Type type, EstimatedStatsAndCost estimate)
             this.estimate = requireNonNull(estimate, "estimate is null");
         }
     }
+
+    private static class ExponentialSleeper
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:7073]: would `assertEventually`, or `FailSafe` do?

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 3]
```diff
@@ -4129,10 +4133,19 @@ public void testParquetTimestampPredicatePushdown()
                 "SELECT * FROM test_parquet_timestamp_predicate_pushdown WHERE t > TIMESTAMP '2012-10-31 01:00'");
         assertEquals(getQueryInfo(queryRunner, queryResult).getQueryStats().getProcessedInputDataSize().toBytes(), 0);
 
-        queryResult = queryRunner.executeWithQueryId(
-                getSession(),
-                "SELECT * FROM test_parquet_timestamp_predicate_pushdown WHERE t = TIMESTAMP '2012-10-31 01:00'");
-        assertThat(getQueryInfo(queryRunner, queryResult).getQueryStats().getProcessedInputDataSize().toBytes()).isGreaterThan(0);
+        // TODO: replace this with a simple query stats check once we find a way to wait until all pending updates to query stats have been applied
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:4136]: cc @dain

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java (hunk 2 above)
Comment (losipiuk) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:7073]: See conversation above
